#Тема «Создание признакового пространства»

##Продолжим обработку данных с Твиттера. 

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform().

 Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
*	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
*	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
*	Исключим стоп-слова с помощью stop_words='english'. 
*	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform().

 Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
*	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
*	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
*	Исключим стоп-слова с помощью stop_words='english'.
*	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

3. Проверьте ваши векторайзеры на корпусе который использовали на вебинаре, составьте таблицу метод векторизации и скор который вы получили (в методах векторизации по изменяйте параметры что бы добиться лучшего скора) обратите внимание как падает/растёт скор при уменьшении количества фичей, и изменении параметров, так же попробуйте применить к векторайзерам PCA для сокращения размерности посмотрите на качество сделайте выводы


In [1]:
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from gensim.models import Word2Vec
import pickle
from sklearn import model_selection, preprocessing, linear_model
from sklearn.metrics import accuracy_score

import warnings
warnings.filterwarnings("ignore")

In [2]:
#Загрузка файла
with open('/content/drive/MyDrive/NLP/Less_01/df_processed.pkl', 'rb') as f:
    df = pickle.load(f)
df

id  label                                              tweet  \
0          1    0.0   @user when a father is dysfunctional and is s...   
1          2    0.0  @user @user thanks for #lyft credit i can't us...   
2          3    0.0                                bihday your majesty   
3          4    0.0  #model   i love u take with u all the time in ...   
4          5    0.0             factsguide: society now    #motivation   
...      ...    ...                                                ...   
49154  49155    NaN  thought factory: left-right polarisation! #tru...   
49155  49156    NaN  feeling like a mermaid 😘 #hairflip #neverre...   
49156  49157    NaN  #hillary #campaigned today in #ohio((omg)) &am...   
49157  49158    NaN  happy, at work conference: right mindset leads...   
49158  49159    NaN  my   song "so glad" free download!  #shoegaze ...   

                                             clean_tweet  \
0      when father is dysfunctional and is so selfish...   
1      thanks for lyft credit cannot use cause they d...   
2                                    bihday your majesty   
3        model love you take with you all the time in ur   
4                      factsguide society now motivation   
...                                                  ...   
49154  thought factory leftright polarisation trump u...   
49155  feeling like mermaid hairflip neverready forma...   
49156  hillary campaigned today in ohioomg amp used w...   
49157  happy at work conference right mindset leads t...   
49158  my song so glad free download shoegaze newmusi...   

                                             tweet_token  \
0      [when, father, is, dysfunctional, and, is, so,...   
1      [thanks, for, lyft, credit, can, not, use, cau...   
2                                [bihday, your, majesty]   
3      [model, love, you, take, with, you, all, the, ...   
4                 [factsguide, society, now, motivation]   
...                                                  ...   
49154  [thought, factory, leftright, polarisation, tr...   
49155  [feeling, like, mermaid, hairflip, neverready,...   
49156  [hillary, campaigned, today, in, ohioomg, amp,...   
49157  [happy, at, work, conference, right, mindset, ...   
49158  [my, song, so, glad, free, download, shoegaze,...   

                                    tweet_token_filtered  \
0      [father, dysfunctional, selfish, drags, kids, ...   
1      [thanks, lyft, credit, use, cause, offer, whee...   
2                                      [bihday, majesty]   
3                          [model, love, take, time, ur]   
4                      [factsguide, society, motivation]   
...                                                  ...   
49154  [thought, factory, leftright, polarisation, tr...   
49155  [feeling, like, mermaid, hairflip, neverready,...   
49156  [hillary, campaigned, today, ohioomg, amp, use...   
49157  [happy, work, conference, right, mindset, lead...   
49158  [song, glad, free, download, shoegaze, newmusi...   

                                           tweet_stemmed  \
0      [father, dysfunct, selfish, drag, kid, dysfunc...   
1      [thank, lyft, credit, use, caus, offer, wheelc...   
2                                      [bihday, majesti]   
3                          [model, love, take, time, ur]   
4                            [factsguid, societi, motiv]   
...                                                  ...   
49154  [thought, factori, leftright, polaris, trump, ...   
49155  [feel, like, mermaid, hairflip, neverreadi, fo...   
49156  [hillari, campaign, today, ohioomg, amp, use, ...   
49157  [happi, work, confer, right, mindset, lead, cu...   
49158  [song, glad, free, download, shoegaz, newmus, ...   

                                        tweet_lemmatized  
0      [father, dysfunct, selfish, drag, kid, dysfunc...  
1      [thank, lyft, credit, use, caus, offer, wheelc...  
2                                      [bihday, majesti]  
3

In [3]:
df = df.dropna()
df

id  label                                              tweet  \
0          1    0.0   @user when a father is dysfunctional and is s...   
1          2    0.0  @user @user thanks for #lyft credit i can't us...   
2          3    0.0                                bihday your majesty   
3          4    0.0  #model   i love u take with u all the time in ...   
4          5    0.0             factsguide: society now    #motivation   
...      ...    ...                                                ...   
31957  31958    0.0  ate @user isz that youuu?😍😍😍😍😍ð...   
31958  31959    0.0    to see nina turner on the airwaves trying to...   
31959  31960    0.0  listening to sad songs on a monday morning otw...   
31960  31961    1.0  @user #sikh #temple vandalised in in #calgary,...   
31961  31962    0.0                   thank you @user for you follow     

                                             clean_tweet  \
0      when father is dysfunctional and is so selfish...   
1      thanks for lyft credit cannot use cause they d...   
2                                    bihday your majesty   
3        model love you take with you all the time in ur   
4                      factsguide society now motivation   
...                                                  ...   
31957                                 ate isz that youuu   
31958  to see nina turner on the airwaves trying to w...   
31959  listening to sad songs on monday morning otw t...   
31960  sikh temple vandalised in in calgary wso conde...   
31961                           thank you for you follow   

                                             tweet_token  \
0      [when, father, is, dysfunctional, and, is, so,...   
1      [thanks, for, lyft, credit, can, not, use, cau...   
2                                [bihday, your, majesty]   
3      [model, love, you, take, with, you, all, the, ...   
4                 [factsguide, society, now, motivation]   
...                                                  ...   
31957                            [ate, isz, that, youuu]   
31958  [to, see, nina, turner, on, the, airwaves, try...   
31959  [listening, to, sad, songs, on, monday, mornin...   
31960  [sikh, temple, vandalised, in, in, calgary, ws...   
31961                     [thank, you, for, you, follow]   

                                    tweet_token_filtered  \
0      [father, dysfunctional, selfish, drags, kids, ...   
1      [thanks, lyft, credit, use, cause, offer, whee...   
2                                      [bihday, majesty]   
3                          [model, love, take, time, ur]   
4                      [factsguide, society, motivation]   
...                                                  ...   
31957                                  [ate, isz, youuu]   
31958  [see, nina, turner, airwaves, trying, wrap, ma...   
31959  [listening, sad, songs, monday, morning, otw, ...   
31960  [sikh, temple, vandalised, calgary, wso, conde...   
31961                                    [thank, follow]   

                                           tweet_stemmed  \
0      [father, dysfunct, selfish, drag, kid, dysfunc...   
1      [thank, lyft, credit, use, caus, offer, wheelc...   
2                                      [bihday, majesti]   
3                          [model, love, take, time, ur]   
4                            [factsguid, societi, motiv]   
...                                                  ...   
31957                                  [ate, isz, youuu]   
31958  [see, nina, turner, airwav, tri, wrap, mantl, ...   
31959  [listen, sad, song, monday, morn, otw, work, sad]   
31960  [sikh, templ, vandalis, calgari, wso, condemn,...   
31961                                    [thank, follow]   

                                        tweet_lemmatized  
0      [father, dysfunct, selfish, drag, kid, dysfunc...  
1      [thank, lyft, credit, use, caus, offer, wheelc...  
2                                      [bihday, majesti]  
3

##Count vectorizer

In [4]:
count_vectorizer = CountVectorizer(max_df=.9, max_features=1000, stop_words='english')
values = count_vectorizer.fit_transform(df['tweet_stemmed'].astype('str'))

feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(values.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
31957    0        0       0        0    0       0      0       0   0      0   
31958    0        0       0        0    0       0      0       0   0      0   
31959    0        0       0        0    0       0      0       0   0      0   
31960    0        0       0        0    1       0      0       0   0      0   
31961    0        0       0        0    0       0      0       0   0      0   

       ...  year  yesterday  yo  yoga  york  young  youth  youtub  yr  yummi  
0      ...     0          0   0     0     0      0      0       0   0      0  
1      ...     0          0   0     0     0      0      0       0   0      0  
2      ...     0          0   0     0     0      0      0       0   0      0  
3      ...     0          0   0     0     0      0      0       0   0      0  
4      ...     0          0   0     0     0      0      0       0   0      0  
...    ...   ...        ...  ..   ...   ...    ...    ...     ...  ..    ...  
31957  ...     0          0   0     0     0      0      0       0   0      0  
31958  ...     0          0   0     0     0      0      0       0   0      0  
31959  ...     0          0   0     0     0      0      0       0   0      0  
31960  ...     0          0   0     0     0      0      0       0   0      0  
31961  ...     0          0   0     0     0      0      0       0   0      0  

[31962 rows x 1000 columns]

In [5]:
count_vectorizer = CountVectorizer(max_df=.9, max_features=1000, stop_words='english')
values = count_vectorizer.fit_transform(df['tweet_lemmatized'].astype('str'))

feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(values.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
31957    0        0       0        0    0       0      0       0   0      0   
31958    0        0       0        0    0       0      0       0   0      0   
31959    0        0       0        0    0       0      0       0   0      0   
31960    0        0       0        0    1       0      0       0   0      0   
31961    0        0       0        0    0       0      0       0   0      0   

       ...  year  yesterday  yo  yoga  york  young  youth  youtub  yr  yummi  
0      ...     0          0   0     0     0      0      0       0   0      0  
1      ...     0          0   0     0     0      0      0       0   0      0  
2      ...     0          0   0     0     0      0      0       0   0      0  
3      ...     0          0   0     0     0      0      0       0   0      0  
4      ...     0          0   0     0     0      0      0       0   0      0  
...    ...   ...        ...  ..   ...   ...    ...    ...     ...  ..    ...  
31957  ...     0          0   0     0     0      0      0       0   0      0  
31958  ...     0          0   0     0     0      0      0       0   0      0  
31959  ...     0          0   0     0     0      0      0       0   0      0  
31960  ...     0          0   0     0     0      0      0       0   0      0  
31961  ...     0          0   0     0     0      0      0       0   0      0  

[31962 rows x 1000 columns]

##TF IDF vectorizer

In [6]:
tfidf_vectorizer = TfidfVectorizer(max_df=.9, max_features=1000, stop_words='english')
values = tfidf_vectorizer.fit_transform(df['tweet_stemmed'].astype('str'))

feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ...    ...   
31957  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
31958  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
31959  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
31960  0.0      0.0     0.0      0.0  1.0     0.0    0.0     0.0  0.0    0.0   
31961  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...  year  yesterday   yo  yoga  york  young  youth  youtub   yr  yummi  
0      ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
1      ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
2      ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
3      ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
4      ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
...    ...   ...        ...  ...   ...   ...    ...    ...     ...  ...    ...  
31957  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
31958  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
31959  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
31960  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  
31961  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0  0.0    0.0  

[31962 rows x 1000 columns]

In [7]:
tfidf_vectorizer = TfidfVectorizer(max_df=.9, max_features=1000, stop_words='english')
values = tfidf_vectorizer.fit_transform(df['tweet_lemmatized'].astype('str'))

feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values.toarray(), columns = feature_names)

abl  absolut  accept  account       act  action  actor  actual   ad  \
0      0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
1      0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
2      0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
3      0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
4      0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
...    ...      ...     ...      ...       ...     ...    ...     ...  ...   
31957  0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
31958  0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
31959  0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   
31960  0.0      0.0     0.0      0.0  0.649536     0.0    0.0     0.0  0.0   
31961  0.0      0.0     0.0      0.0  0.000000     0.0    0.0     0.0  0.0   

       adapt  ...  year  yesterday   yo  yoga  york  young  youth  youtub  \
0        0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
1        0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
2        0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
3        0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
4        0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
...      ...  ...   ...        ...  ...   ...   ...    ...    ...     ...   
31957    0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
31958    0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
31959    0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
31960    0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   
31961    0.0  ...   0.0        0.0  0.0   0.0   0.0    0.0    0.0     0.0   

        yr  yummi  
0      0.0    0.0  
1      0.0    0.0  
2      0.0    0.0  
3      0.0    0.0  
4      0.0    0.0  
...    ...    ...  
31957  0.0    0.0  
31958  0.0    0.0  
31959  0.0    0.0  
31960  0.0    0.0  
31961  0.0    0.0  

[31962 rows x 1000 columns]

In [9]:
df.columns

Index(['id', 'label', 'tweet', 'clean_tweet', 'tweet_token',
       'tweet_token_filtered', 'tweet_stemmed', 'tweet_lemmatized'],
      dtype='object')

In [14]:
df['label'].value_counts()

0.0    29720
1.0     2242
Name: label, dtype: int64

#Учитывая такой перекос моделька скоре всего будет переобучаться.

In [31]:
col = ['tweet_stemmed', 'tweet_lemmatized']
print('CountVectorizer')
print('-' *50)
for i in col:
  train_x, valid_x, train_y, valid_y = model_selection.train_test_split(df[i], df['label'])
  count_vectorizer = CountVectorizer(max_df=.9, max_features=1000, stop_words='english')
  xtrain_count =  count_vectorizer.fit_transform(train_x.astype('str'))
  xvalid_count =  count_vectorizer.fit_transform(valid_x.astype('str'))
  classifier = linear_model.LogisticRegression()
  classifier.fit(xtrain_count, train_y)
  predictions = classifier.predict(xvalid_count)
  acc = accuracy_score(valid_y, predictions)
  print(i, ' - ', acc)
print('-' *50)
print('TfidfVectorizer')
print('-' *50)
for i in col:
  train_x, valid_x, train_y, valid_y = model_selection.train_test_split(df[i], df['label'])
  tfidf_vectorizer = TfidfVectorizer(max_df=.9, max_features=1000, stop_words='english')
  xtrain_tfidf = tfidf_vectorizer.fit_transform(train_x.astype('str'))
  xvalid_tfidf = tfidf_vectorizer.fit_transform(valid_x.astype('str'))
  classifier = linear_model.LogisticRegression()
  classifier.fit(xtrain_tfidf, train_y)
  predictions = classifier.predict(xvalid_tfidf)
  acc = accuracy_score(valid_y, predictions)
  print(i, ' - ', acc)

CountVectorizer
--------------------------------------------------
tweet_stemmed  -  0.9169065198348142
tweet_lemmatized  -  0.9017644850456764
--------------------------------------------------
TfidfVectorizer
--------------------------------------------------
tweet_stemmed  -  0.9256663746715055
tweet_lemmatized  -  0.9240395444875485


#Подкручивать параметры скорее всего не получиться потому что показатели уже на уровне переобучения (скорее запоминания)

* 0.0  -  29720
* 1.0   -  2242
